src/main/kotlin/calculator/domain/Calculator.kt

```diff
@@ -0,0 +1,35 @@
+package calculator.domain
+
+import calculator.extenstion.isNotInt
+
+private const val EMPTY_OPERATOR = ""
+
+object Calculator {
+    tailrec fun calculate(expression: MutableList<String>, accumulator: Int = expression.removeFirst().toInt(), operator: String = EMPTY_OPERATOR): Int {
+        if (expression.isEmpty()) {
+            return accumulator
+        }
+
+        return when {
+            expression.first().isNotInt() -> calculate(expression = expression, accumulator = accumulator, operator = expression.removeFirst())
+            else -> calculate(expression = expression, accumulator = Operator.calculate(accumulator, operator, expression.removeFirst().toInt()), operator = EMPTY_OPERATOR)
+        }
+    }
+}
+
+enum class Operator(
+    private val operator: String,
+    private val calculator: (Int, Int) -> Int
+) {
+    PLUS("+", { num1, num2 -> num1 + num2 }),
+    MINUS("-", { num1, num2 -> num1 - num2 }),
+    MULTIPLY("*", { num1, num2 -> num1 * num2 }),
+    DIVIDE("/", { num1, num2 -> num1 / num2 });
+
+    companion object {
+        fun calculate(num1: Int, operator: String, num2: Int): Int {
+            return values().firstOrNull { it.operator == operator }?.calculator?.invoke(num1, num2)
+                ?: throw IllegalArgumentException("$operator is invalid operator")
+        }
```

Comment:
없으면 먼저 `Exception`을 발생시키고, 그 뒤의 로직을 수행하는건 어떨까요?
그러면 `?.` 등도 많이 없어지면서 코드의 가독성은 높아질꺼에요.

```suggestion
        fun calculate(num1: Int, operator: String, num2: Int): Int {
            val calculator = values().firstOrNull { it.operator == operator }
                ?: throw IllegalArgumentException("$operator is invalid operator")

            return calculator.calculator(num1, num2)
        }
```

Reply:
좋은 것 같습니다!
제가 작성한 코드는 체이닝이 많이 생기다보니 가독성이 떨어지네요
해당 부분도 수정하도록 할게요!